File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/usecases/GiftCardIssuingTest.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.commands.IssueGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.giftcard.infrastructure.SqliteGiftCardRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class GiftCardIssuingTest {
    private GiftCardRepository repository;
    private GiftCardIssuing issuance;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("giftcard");
        repository = new SqliteGiftCardRepository(dataSource);
        issuance = new GiftCardIssuing(repository);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new GiftCardIssuing(null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullCommand() {
            assertThatThrownBy(() -> issuance.invoke(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldPersist() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);

            issuance.invoke(new IssueGiftCard(cardId));

            Optional<GiftCard> loaded = repository.findById(cardId);

            assertThat(loaded).isPresent();
            assertThat(loaded.get().id()).isEqualTo(cardId);
            assertThat(loaded.get().balance()).isEqualTo(Money.zero());
        }
    }
}

File: typescript/tests/giftcard/application/usecases/giftCardIssuing.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { IssueGiftCard, issueGiftCard } from '@/giftcard/application/commands/issueGiftCard.js';
import { GiftCardIssuing } from '@/giftcard/application/usecases/giftCardIssuing.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';

describe('GiftCardIssuing', () => {
  let repository: SqliteGiftCardRepository;
  let issuance: GiftCardIssuing;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('giftcard');
    repository = new SqliteGiftCardRepository(database);
    issuance = new GiftCardIssuing(repository);
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new GiftCardIssuing(null as unknown as SqliteGiftCardRepository)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null command', () => {
      expect(() => issuance.invoke(null as unknown as IssueGiftCard)).toThrow();
    });

    it('should persist', () => {
      const cardId = generateId((value) => new GiftCardId(value));

      issuance.invoke(issueGiftCard(cardId));

      const loaded = repository.findById(cardId);
      expect(loaded).not.toBeNull();
      expect(loaded!.id()).toEqual(cardId);
      expect(loaded!.balance()).toEqual(Money.zero());
    });
  });
});
